### PR TITLE
blockstore: add merkle root to ErasureMeta column family

### PR DESCRIPTION
**Pull request by AshwinSekar**
#### Problem
In order to detect duplicate blocks with conflicting merkle roots we need to be able to access the merkle root for each FEC set.

We store the merkle root along with the existing `ErasureMeta` for each FEC set.

#### Summary of Changes
This change allows us to read blockstore in both formats, with or without the merkle root. It is intended to be backported.

Contributes to #33644

### CODE REVIEW

File: ledger/src/blockstore_meta.rs [hunk 1]
```diff
@@ -331,11 +345,12 @@ impl ErasureMeta {
                     num_coding: usize::from(shred.num_coding_shreds().ok()?),
                 };
                 let first_coding_index = u64::from(shred.first_coding_index()?);
-                let erasure_meta = ErasureMeta {
+                let merkle_root = Hash::default();
```
Comment (AshwinSekar) [inline ledger/src/blockstore_meta.rs:348]: next change in master will populate this from the coding shred, but I don't think that needs to be part of this backport change.

File: ledger/src/blockstore_db.rs [hunk 2]
```diff
@@ -802,6 +802,12 @@ pub trait ColumnIndexDeprecation: Column {
     }
 }
 
+/// Helper trait to transition a column between bincode formats
+pub trait BincodeTypeTransition: Column + TypedColumn {
+    /// This should have a different serialized size than the TypedColumn::Type
+    type OldType: Serialize + DeserializeOwned + Into<Self::Type>;
+}
+
```
Comment (behzadnouri) [inline ledger/src/blockstore_db.rs:810]: This type trait and its associated code is pretty verbose and actually adding more complexity here.

Might be simple just to do something like below:
```diff
diff --git a/ledger/src/blockstore_db.rs b/ledger/src/blockstore_db.rs
index b65df82ee0..eebf163400 100644
--- a/ledger/src/blockstore_db.rs
+++ b/ledger/src/blockstore_db.rs
@@ -1586,7 +1586,7 @@ where
     }
 }
 
-impl<C> LedgerColumn<C>
+impl<C: 'static> LedgerColumn<C>
 where
     C: TypedColumn + ColumnName,
 {
@@ -1635,7 +1635,12 @@ where
             &self.read_perf_status,
         );
         if let Some(pinnable_slice) = self.backend.get_pinned_cf(self.handle(), key)? {
-            let value = deserialize(pinnable_slice.as_ref())?;
+            use std::any::TypeId;
+            let value = deserialize(pinnable_slice.as_ref()).or_else(|err| {
+                if TypeId::of::<C>() == TypeId::of::<columns::MerkleErasureMeta>() {
+                    // retry with the old erasure meta type.
+                    // ...
+                }
+                Err(err)
+            })?;
             result = Ok(Some(value))
         }
 
```

File: ledger/src/blockstore_meta.rs [hunk 3]
```diff
@@ -118,24 +118,38 @@ pub struct ShredIndex {
...
     #[serde(rename = "size")]
-    __unused_size: usize,
+    pub(crate) __unused_size: usize,
+    /// Erasure configuration for this erasure set
+    pub(crate) config: ErasureConfig,
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, Serialize, Eq, PartialEq)]
+/// Erasure coding information for merkle shreds
+pub struct MerkleErasureMeta {
```
[the remark this answers is not in the document]
Comment (AshwinSekar) [reply ledger/src/blockstore_meta.rs:138]: yeah good point, was planning to scan using the `first_coding_index`, but we can just put the coding shred index and save some blockstore reads

File: ledger/src/blockstore_meta.rs [hunk 4]
```diff
@@ -118,24 +118,38 @@ pub struct ShredIndex {
...
+    /// Merkle root for this FEC set
+    merkle_root: Hash,
 }
 
 #[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
 pub(crate) struct ErasureConfig {
-    num_data: usize,
-    num_coding: usize,
+    pub(crate) num_data: usize,
+    pub(crate) num_coding: usize,
```
[the remark this answers is not in the document]
Comment (AshwinSekar) [reply ledger/src/blockstore_meta.rs]: yeah for tests, i just made a test impl instead.

File: ledger/src/blockstore_db.rs (hunk 2 above)
Comment (AshwinSekar) [reply ledger/src/blockstore_db.rs:810]: I added it to provide an easy way to transition types of already existing columns in blockstore. Similar to ColumnIndexDeprecation.

All you have to do is implement the trait and TypedColumn. And after the old type is no longer in blockstore, you can just remove the trait impl and the associated migration test.

Figured its worth adding the complexity here to make future transitions frictionless.

**Comment by AshwinSekar** [comment]
After some offline discussion, it seems possible to detect merkle root conflicts without the requirement of an `ErasureMeta`, which we originally thought was not possible. 

Given this, I am closing this in favor of the original idea of adding a new column. We can add a new `MerkleRootMeta` column that contains the merkle root and meta information instead of reusing the `ErasureMeta`. This will avoid the need for any backports as well. 

will continue here #33979

**Pull request closed without merging**